File: searching-sorting-algo/01-searching/02-binary-search.c
#include<stdio.h>
#include<stdlib.h>

/* 
    ->> binary search
    - it is fast searching algorithm with time complexity O(log n)
    - it is based on divide and conquer
    - data item should be in sorted manner
*/

int binarySearch(int array[], int key, int low, int high){
    while(low <= high){
        int mid = (low + high) / 2;

        if(array[mid] == key)
            return 1;
        
        // go to right side
        if(array[mid] < key)
            low = mid + 1;
        // go to left side
        else
            high = mid - 1;
        
    }

    return -1;
}


int main(){

    // sorted array
    int ARRAY[] = {10, 20, 30, 40, 50};
    // length of array 
    int n = sizeof(ARRAY) / sizeof(ARRAY[0]);
    // element to search
    int x = 40;

    int low = 0;
    int high = n-1;

    int result = binarySearch(ARRAY, x, low, high);

    (result == 1)? printf("Element Found"): printf("Element Not Found");

    return 0;
}

File: searching-sorting-algo/01-searching/inJava/02-binary-search.java

class Example{
    public static void main(String[] args){
        // binary search (works only on sorted array)
        int[] ARRAY  = {2, 10, 34, 56, 76};
        int x = 34;

        binarySearch(ARRAY, x);
    }

    public static void binarySearch(int array[], int x){
        int low = 0;
        int high = array.length-1;
        int mid = (low + high) / 2;

        while(low <= high){
            
            if(x == array[mid]){
                System.out.println("Element Found");
                return;
            }

            if(x > array[mid]){
                low = mid+1;
                mid = (low + high)/2;
            }else{
                high = mid - 1;
                mid = (low + high)/2;
            }

        }

        System.out.println("Element Not Found");
        
    }
}